runtime/sema/check_conditions.go

```diff
@@ -40,15 +43,157 @@ func (checker *Checker) visitConditions(conditions []*ast.Condition) {
 	}
 }
 
-func (checker *Checker) checkCondition(condition *ast.Condition) Type {
+func (checker *Checker) checkCondition(condition ast.Condition) Type {
 
-	// check test expression is boolean
-	checker.VisitExpression(condition.Test, BoolType)
+	switch condition := condition.(type) {
+	case *ast.TestCondition:
 
-	// check message expression results in a string
-	if condition.Message != nil {
-		checker.VisitExpression(condition.Message, StringType)
+		// check test expression is boolean
+		checker.VisitExpression(condition.Test, BoolType)
+
+		// check message expression results in a string
+		if condition.Message != nil {
+			checker.VisitExpression(condition.Message, StringType)
+		}
+
+	case *ast.EmitCondition:
+		checker.VisitEmitStatement((*ast.EmitStatement)(condition))
+
+	default:
+		panic(errors.NewUnreachableError())
 	}
 
 	return nil
 }
+
+func (checker *Checker) rewritePostConditions(postConditions ast.Conditions) PostConditionsRewrite {
```

Comment:
Moved from https://github.com/onflow/cadence/pull/2589/files#diff-74d0954852bf407e9259a7c9bb55e8af093f080f00f5e35689c250886c29860aL1961 and refactored into sub-functions